cli/flox-rust-sdk/src/providers/catalog.rs

```diff
@@ -1699,12 +1703,115 @@ impl BaseCatalogInfo {
+
+            info!(%url, %stability, "using page from user provided stability");
+            url
+        },
+        (None, Some(toplevel_derived_url)) => {
+            info!(url=%toplevel_derived_url, "using nixpkgs derived from toplevel group");
+            toplevel_derived_url.clone()
+        },
+        (None, None) => {
+            let base_catalog_info = base_catalog_info_fut.await?;
+
+            let make_error_message = || {
+                let available_stabilities = base_catalog_info.available_stabilities().join(", ");
+                formatdoc! {"
+                    The default stability {} does not exist (or has not yet been populated).
+                    Available stabilities are: {available_stabilities}
+                ", BaseCatalogInfo::DEFAULT_STABILITY}
+            };
+
+            let url = base_catalog_info
+                .url_for_latest_page_with_default_stability()
+                .ok_or_else(|| CatalogClientError::StabilityError(make_error_message()))?;
+
+            info!(%url, "using page from default stability");
+            url
+        },
+    };
+    Ok(url)
+}
+
+/// Returns the nixpkgs URL used for builds and publishes.
+pub async fn get_base_nixpkgs_url(
+    flox: &Flox,
+    stability: Option<&str>,
+    env_metadata: &CheckedEnvironmentMetadata,
+) -> Result<BaseCatalogUrl, CatalogClientError> {
+    let base_catalog_info_fut = flox.catalog_client.get_base_catalog_info();
+
+    base_catalog_url_for_stability_arg(
+        stability,
+        base_catalog_info_fut,
+        env_metadata.toplevel_catalog_ref.as_ref(),
+    )
+    .await
+}
+
 pub mod test_helpers {
+    use tempfile::TempDir;
+
     use super::*;
+    use crate::flox::test_helpers::create_test_token;
+    use crate::flox::{Flox, FloxhubToken};
+    use crate::providers::auth::{Auth, AuthProvider};
 
     pub static UNIT_TEST_GENERATED: LazyLock<PathBuf> =
         LazyLock::new(|| PathBuf::from(std::env::var("UNIT_TEST_GENERATED").unwrap()));
 
+    /// Whether to record mock data and in which situations.
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+    enum RecordMockData {
+        /// Only record new mock data if it's missing.
+        Missing,
+        /// Don't record new mock data.
+        #[default]
+        False,
+        /// Re-record all mock data.
+        Force,
+    }
+
+    /// Returns in which circumstances mock data should be recorded based on
+    /// the value of the `_FLOX_UNIT_TEST_RECORD` environment variable.
+    ///
+    /// Values of "missing", "true", or "1" will generate a recording for a
+    /// missing mock. An unset variable or a value of "false" will only replay
+    /// existing recordings. The value "force" will unconditionally regenerate
+    /// mock data. Any other value will cause a panic.
+    fn get_record_directive() -> RecordMockData {
+        let s = std::env::var("_FLOX_UNIT_TEST_RECORD").unwrap_or_default();
+        match s.as_str() {
```

Comment:
Reviewer note: I'm being lenient here with which values are accepted, since it's common for `true` and `1` to be roughly equivalent in environment variables. I extend that logic to the other variants as well just for consistency.